README
Name
    File::MealMaster - OO access to the MealMaster recipe files

Version
    This documents version v0.18.$Rev: 1 $ of File::MealMaster

Synopsis
       use File::MealMaster;

       my $mealmaster_ref = File::MealMaster->new( $mealmaster_attributes );

Description
    Extends File::DataClass::Schema. Defines the "recipes" result source

Configuration and Environment
    Defines these attributes:

    "cache_attributes"
    "cache_class"
    "result_source_attributes"
    "source_name"
       Defaults to "recipes". The result source name in the schema
       definition

    Modifies these methods;

    "resultset"
    "source"

Subroutines/Methods
  make_key
    Exposes make_key

Diagnostics
    None

Dependencies
    File::DataClass::Schema
    File::MealMaster::Result
    File::MealMaster::Storage
    Moo

Incompatibilities
    There are no known incompatibilities in this module

Bugs and Limitations
    There are no known bugs in this module. Please report problems to the
    address below. Patches are welcome

Author
    Peter Flanigan, "<pjfl@cpan.org>"

License and Copyright
    Copyright (c) 2013 Peter Flanigan. All rights reserved

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself. See perlartistic

    This program is distributed in the hope that it will be useful, but
    WITHOUT WARRANTY; without even the implied warranty of MERCHANTABILITY
    or FITNESS FOR A PARTICULAR PURPOSE



Changes
Revision history for File::MealMaster

0.18.1    2013-08-07 11:02:06
          - Raised minimum Perl version to v5.10.1

0.17.9    2013-07-30 17:47:26
          - New method for CPANTesting min Perl ver
          - Added fremen and frogman to tier gruber
          - Aborting Bingoes failed see df276fba-f57c-11e2-8c80-50d7c5c10595
            No choice but to send this smoker to sleep 10 while 1.
            Chris Williams, put my pause id in your stop list
          - Bumped File::DataClass dependency
          - Mooified

0.16.5    2013-04-16 12:24:30
          - Added missing dependency on parent
          - Updated POD
          - Added t/critic.rc and fixed case of emacs eval line
          - Extracted code and tests from File::DataClass distribution

0.1.1     2013-04-10 22:51:57
          - Created initial project files



inc/CPANTesting.pm
# @(#)Ident: CPANTesting.pm 2013-07-30 16:46 pjf ;

package CPANTesting;

use strict;
use warnings;

use Sys::Hostname; my $host = lc hostname; my $osname = lc $^O;

# Is this an attempted install on a CPAN testing platform?
sub is_testing { !! ($ENV{AUTOMATED_TESTING} || $ENV{PERL_CR_SMOKER_CURRENT}
                 || ($ENV{PERL5OPT} || q()) =~ m{ CPAN-Reporter }mx) }

sub should_abort {
   is_testing() or return 0;

   $host eq q(xphvmfred) and return
      "ABORT: ${host} - cc06993e-a5e9-11e2-83b7-87183f85d660";
   return 0;
}

sub test_exceptions {
   my $p = shift; my $perl_ver = $p->{requires}->{perl};

   is_testing()         or  return 0;
   $] < $perl_ver       and return "TESTS: Perl minimum ${perl_ver}";
   $p->{stop_tests}     and return 'TESTS: CPAN Testing stopped in Build.PL';
   $osname eq q(mirbsd) and return 'TESTS: Mirbsd OS unsupported';
#  $host   eq q(broken) and return "tests: <CPAN Testing uuid>";
   return 0;
}

1;

__END__


lib/File/MealMaster.pm
# @(#)$Ident: MealMaster.pm 2013-08-07 00:37 pjf ;

package File::MealMaster;

use 5.010001;
use namespace::sweep;
use version; our $VERSION = qv( sprintf '0.18.%d', q$Rev: 1 $ =~ /\d+/gmx );

use File::DataClass::Constants;
use File::DataClass::Types  qw( Str );
use File::MealMaster::Result;
use Moo;

extends q(File::DataClass::Schema);

has '+cache_attributes' => default => sub {
   (my $ns = lc __PACKAGE__) =~ s{ :: }{-}gmx; return { namespace => $ns, }
};

has '+cache_class' => default => 'none';

has '+result_source_attributes' => default => sub {
   { recipes          => {
      attributes      => [ qw(categories directions
                              ingredients title yield) ],
      defaults        => { categories => [], ingredients => [] },
      resultset_attributes => {
         result_class => q(File::MealMaster::Result), }, }, }
};

has '+storage_class' => default => '+File::MealMaster::Storage';

has 'source_name' => is => 'ro', isa => Str, default => 'recipes';

around 'source' => sub {
   my ($orig, $self) = @_; return $self->$orig( $self->source_name );
};

around 'resultset' => sub {
   my ($orig, $self) = @_; return $self->$orig( $self->source_name );
};

sub make_key {
   my ($self, $title) = @_; return $self->source->storage->make_key( $title );
}


1;

__END__

=pod

=encoding utf8

=head1 Name

File::MealMaster - OO access to the MealMaster recipe files

=head1 Version

This documents version v0.18.$Rev: 1 $ of L<File::MealMaster>

=head1 Synopsis

   use File::MealMaster;

   my $mealmaster_ref = File::MealMaster->new( $mealmaster_attributes );

=head1 Description

Extends L<File::DataClass::Schema>. Defines the C<recipes> result source

=head1 Configuration and Environment

Defines these attributes:

=over 3

=item C<cache_attributes>

=item C<cache_class>

=item C<result_source_attributes>

=item C<source_name>

Defaults to C<recipes>. The result source name in the schema definition

=back

Modifies these methods;

=over 3

=item C<resultset>

=item C<source>

=back

=head1 Subroutines/Methods

=head2 make_key

Exposes L<make_key|File::MealMaster::Storage/make_key>

=head1 Diagnostics

None

=head1 Dependencies

=over 3

=item L<File::DataClass::Schema>

=item L<File::MealMaster::Result>

=item L<File::MealMaster::Storage>

=item L<Moo>

=back

=head1 Incompatibilities

There are no known incompatibilities in this module

=head1 Bugs and Limitations

There are no known bugs in this module.
Please report problems to the address below.
Patches are welcome

=head1 Author

Peter Flanigan, C<< <pjfl@cpan.org> >>

=head1 License and Copyright

Copyright (c) 2013 Peter Flanigan. All rights reserved

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself. See L<perlartistic>

This program is distributed in the hope that it will be useful,
but WITHOUT WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE

=cut

# Local Variables:
# mode: perl
# tab-width: 3
# End:


lib/File/MealMaster/Result.pm
# @(#)$Ident: Result.pm 2013-06-22 19:26 pjf ;

package File::MealMaster::Result;

use namespace::clean -except => 'meta';
use version; our $VERSION = qv( sprintf '0.18.%d', q$Rev: 1 $ =~ /\d+/gmx );

use English                 qw( -no_match_vars );
use File::DataClass::Constants;
use File::DataClass::Types  qw( Str );
use Moo;

extends qw(File::DataClass::Result);

my $DATA = do { local $RS = undef; <DATA> };

has '_render_template' => is => 'rw', isa => Str, default => $DATA;

sub render {
   my $self          = shift;
   my $template      = $self->_storage->template;
   my $template_data = $self->_render_template;
   my $buffer        = NUL;

   $template->process( \$template_data, $self, \$buffer )
      or $buffer = $template->error;

   return $buffer;
}

1;

=pod

=head1 Name

File::MealMaster::Result - MealMaster food recipes custom methods

=head1 Version

0.16.$Rev: 1 $

=head1 Synopsis

=head1 Description

Adds custom methods to the recipe result class

=head1 Configuration and Environment

=head1 Subroutines/Methods

=head2 render

Return the HTML representation of this recipe. Uses the internal template
in the data block at the end of this file

=head1 Diagnostics

None

=head1 Dependencies

=over 3

=item L<Data::Section>

=item L<File::DataClass::Result>

=back

=head1 Incompatibilities

There are no known incompatibilities in this module

=head1 Bugs and Limitations

There are no known bugs in this module.
Please report problems to the address below.
Patches are welcome

=head1 Author

Peter Flanigan, C<< <Support at RoxSoft.co.uk> >>

=head1 License and Copyright

Copyright (c) 2013 Peter Flanigan. All rights reserved

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself. See L<perlartistic>

This program is distributed in the hope that it will be useful,
but WITHOUT WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE

=cut

# Local Variables:
# mode: perl
# tab-width: 3
# End:

__DATA__
<table class="recipe">
   <tr><td class="label">Title</td><td>[% title | html %]</td></tr>
   <tr><td class="label">Categories</td><td>
	   [% FOREACH category IN categories %][% category | html %] | [% END %]
   </td></tr>
   <tr><td class="label">Yield</td><td>[% yield | html %]</td></tr>
	<tr><td class="label">Ingredients</td><td><table class="recipe">
   [% FOREACH ingredient IN ingredients %]
     <tr><td class="nowrap" width="1%">[% ingredient.quantity | html %]</td>
         <td width="1%">[% ingredient.measure | html %]</td>
        <td>[% ingredient.product | html %]</td></tr>
   [% END %]</table></td></tr>
   <tr><td class="label">Directions</td>
     <td>[% directions | html | html_line_break %]</td></tr>
</table>


lib/File/MealMaster/Storage.pm
# @(#)$Ident: Storage.pm 2013-06-22 19:31 pjf ;

package File::MealMaster::Storage;

use namespace::clean -except => 'meta';
use version; our $VERSION = qv( sprintf '0.18.%d', q$Rev: 1 $ =~ /\d+/gmx );

use English                    qw( -no_match_vars );
use File::DataClass::Constants;
use File::DataClass::Functions qw( throw );
use File::DataClass::Types     qw( Object Str );
use Moo;
use MooX::Augment -class;
use Template;
use Template::Stash;

extends q(File::DataClass::Storage);

my $DATA = do { local $RS = undef; <DATA> };

has '+extn'          => default => '.mmf';

has 'template'       => is => 'lazy', isa => Object;

has 'write_template' => is => 'ro',   isa => Str, default => $DATA;

augment '_read_file' => sub {
   my ($self, $rdr) = @_;

   $self->encoding and $rdr->encoding( $self->encoding );

   return $rdr->all;
};

around '_read_file' => sub {
   my ($orig, $self, @args) = @_;

   my ($data, $mtime) = $self->$orig( @args ); my $order = 0; my $recipes = {};

   for my $recipe (MealMasterMashup->new()->parse( $data )) {
      $recipe->{ _order_by } = $order++;
      $recipes->{ $self->make_key( $recipe->title ) } = $recipe;
   }

   return ({ $self->schema->source_name => $recipes }, $mtime);
};

augment '_write_file' => sub {
   my ($self, $wtr, $data) = @_;

   $self->encoding and $wtr->encoding( $self->encoding );

   return $self->_write_filter( $wtr, $data );
};

sub make_key {
   my ($self, $title) = @_; my $key = $title || NUL;

   $key =~ s{ [^a-zA-Z0-9 ] }{}gmsx;
   $key =~ s{ \s+ }{ }gmsx;
   $key =~ s{ \A \s+ }{}msx;
   $key =~ s{ \s+ \z }{}msx;
   $key =  join q(_), map { ucfirst $_ } split SPC, lc $key;

   return $key;
}

# Private methods
sub _write_filter {
   my ($self, $wtr, $data) = @_; $data ||= {};

   my $recipes       = $data->{ $self->schema->source_name } || {};
   my $template_data = $self->write_template;
   my $output        = NUL;

   for (sort { __original_order( $recipes, $a, $b ) } keys %{ $recipes }) {
      my $buffer = NUL;

      $self->template->process( \$template_data, $recipes->{ $_ }, \$buffer )
         or throw $self->template->error;
      $output .= $buffer;
   }

   $wtr->println( $output );
   return $data;
}

# Private methods
sub _build_template {
   my $self = shift;
   my $args = { INTERPOLATE => FALSE, COMPILE_DIR => $self->schema->tempdir };
   my $new  = Template->new( $args ) or throw Template->error;

   $Template::Stash::SCALAR_OPS->{sprintf} = sub {
      my ($val, $format) = @_; return sprintf $format, $val;
   };

   return $new;
}

# Private subroutines
sub __original_order {
   my ($hash, $lhs, $rhs) = @_;

   # New elements will be  added at the end
   exists $hash->{ $lhs }->{_order_by} or return  1;
   exists $hash->{ $rhs }->{_order_by} or return -1;
   return $hash->{ $lhs }->{_order_by} <=> $hash->{ $rhs }->{_order_by};
}

package # Hide from indexer
   MealMasterMashup;

use parent 'MealMaster';

sub parse {
   # Copyright (C) 2005, Leon Brocard
   # Needed a version that takes scalar data
   # Also patched to handle whitespace better
   my ($self, $data) = @_;

   $data and $data =~ /^(MMMMM|-----).+Meal-Master/ or return;

   my @parts = split /^(?:MMMMM|-----).+Meal-Master.+$/m, $data;
   my @recipes;

   foreach my $part (@parts) {
      $part =~ s/^\s+//;
      my $recipe = MealMaster::Recipe->new;
      my $lines  = [ split /\n/, $part ];

      my $line;

      while (1) {
         $line = $self->_newline($lines);
         last unless defined $line;
         last if $line =~ /Title:/;
      }
      next unless defined $line;

      my ($title) = $line =~ /Title: (.+)$/;
      next unless $title;
      $title =~ s/^ +//;
      $title =~ s/ +$//;
      $recipe->title($title);

      $line = $self->_newline($lines);
      my ($categories) = $line =~ /Categories: (.+)$/;

      my @categories;
      @categories = split ', ', $categories if $categories;
      $recipe->categories(\@categories);

      $line = $self->_newline($lines);
      my ($yield) = $line =~ /(?:Yield|Servings): +(.+)$/;
      next unless $yield;
      $recipe->yield($yield);

      my $dflag = 0;
      my $ingredients;
      my $directions;

      while (defined($line = $self->_newline($lines))) {
         next unless $line;

         last if (!defined $line);
         next if (($dflag == 0) && ($line =~ m|^\s*$|));

         if ($line =~ /^[M-]+$/) {
            last;
         } elsif ($line =~ m(^[M|-]{4,})) {
            $line =~ s|^MMMMM||;
            $line =~ s|^\-+||;
            $line =~ s|\-+$||;
            $line =~ s|^ +||;
            $line =~ s|:$||;
            $directions .= "$line\n";
         } elsif ($line =~ m/^ *([A-Z ]+):$/) {
            $line =~ s|^\s+||;
            $directions .= "$line\n";
         } elsif (length($line) > 12
                  && (substr($line, 0, 7) =~ m|^[ 0-9\.\/]+$|)
                  && (substr($line, 7, 4) =~ m|^ .. $|))
         {
            $ingredients .= "$line\n";
         } else {
            if ($line) {
               $line =~ s|^\s+||;
               $directions .= "$line\n";
               $dflag = 1;
            }
         }
      }

      $ingredients = $self->_parse_ingredients($ingredients);

      $recipe->ingredients($ingredients);
      $recipe->directions($directions);

      push @recipes, $recipe;
   }
   return @recipes;
}

package File::MealMaster::Storage;

1;

=pod

=head1 Name

File::MealMaster::Storage - MealMaster food recipe file storage

=head1 Version

0.16.$Rev: 1 $

=head1 Synopsis

=head1 Description

=head1 Configuration and Environment

Defines the following attributes;

=over 3

=item C<extn>

=item C<write_template>

=back

=head1 Subroutines/Methods

=head2 make_key

=head1 Diagnostics

None

=head1 Dependencies

=over 3

=item L<File::DataClass::Storage>

=back

=head1 Incompatibilities

There are no known incompatibilities in this module

=head1 Bugs and Limitations

There are no known bugs in this module.
Please report problems to the address below.
Patches are welcome

=head1 Author

Peter Flanigan, C<< <Support at RoxSoft.co.uk> >>

=head1 License and Copyright

Copyright (c) 2013 Peter Flanigan. All rights reserved

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself. See L<perlartistic>

This program is distributed in the hope that it will be useful,
but WITHOUT WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE

=cut

# Local Variables:
# mode: perl
# tab-width: 3
# End:

__DATA__
MMMMM----- Recipe via Meal-Master (tm) v8.05
 
      Title: [% title %]
 Categories: [% categories.join(', ') %]
      Yield: [% yield %]
 
[% FOREACH ingredient IN ingredients -%]
[% ingredient.quantity.sprintf('%7.7s') -%]
 [% ingredient.measure.sprintf('%-2.2s') -%]
 [% ingredient.product.sprintf('%-.29s') %]
[% END -%]
 
  [% directions.split('\n').join("\n  ") %]
 
MMMMM
 


Build.PL
# @(#)$Ident: Build.PL 2013-08-07 10:00 pjf ;

use strict;
use warnings;
use inc::Bob; # Module::Build subclass
use version; our $VERSION = qv( sprintf '0.18.%d', q$Rev: 1 $ =~ /\d+/gmx );

Bob->new( {
   module    => 'File::MealMaster',
   home_page => 'http://www.roxsoft.co.uk',
   requires  => {
      'perl'             => 5.010001,
      'namespace::clean' => 0.23,
      'namespace::sweep' => 0.006,
      'parent'           => 0.224,
      'Class::Method::ModifiersX::Augment' => 0.001,
      'File::DataClass'  => q(0.22.0),
      'MealMaster'       => 0.28,
      'Moo'              => 1.002,
      'Template'         => 2.22, },
   recommends            => {
      'App::cpanminus'   => 1.6005, },
   configure_requires    => {
      'version'          => 0.88,
      'Module::Build'    => 0.4004, },
   build_requires        => {
      'version'          => 0.88,
      'Module::Build'    => 0.4004,
      'Text::Diff'       => 1.37, },
} )->create_build_script; # Goal!

# Local Variables:
# eval: (load-project-state "file-mealmaster")
# mode: perl
# tab-title: File::MealMaster
# tab-width: 3
# End:
